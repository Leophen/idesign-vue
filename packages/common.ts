import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
