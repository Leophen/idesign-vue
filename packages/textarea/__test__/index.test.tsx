import { vi } from 'vitest';
import { Textarea } from '../index';
import { render } from '../../common'

describe('Textarea 组件测试', () => {
  const renderTextarea = (textarea) => {
    const wrapper = render(textarea);
    return wrapper.find('.i-textarea__inner')
  }

  it('create', () => {
    const wrapper = render(<Textarea />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('placeholder', () => {
    const wrapper = render(<Textarea placeholder='请输入内容' />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('value', () => {
    const wrapper = render(<Textarea modelValue='iDesign' />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('defaultValue', () => {
    const wrapper = render(<Textarea defaultValue='iDesign' />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('onChange', () => {
    const changeFn = vi.fn();
    renderTextarea(<Textarea onChange={changeFn} />).trigger('change');
    expect(changeFn).toHaveBeenCalled();
  });

  it('disabled', () => {
    const changeFn = vi.fn();
    renderTextarea(<Textarea onChange={changeFn} disabled />).trigger('change');
    expect(renderTextarea(<Textarea onChange={changeFn} disabled />).classes()).toContain('i-textarea__inner-is-disabled');
    expect(changeFn).toBeCalledTimes(0);
  });

  it('readonly', () => {
    expect(renderTextarea(<Textarea readonly />).classes()).toContain('i-textarea__inner-is-readonly');
  });

  it('maxLength', () => {
    const wrapper = render(<Textarea maxLength={15} />);
    expect(wrapper.find('.i-textarea--limit').exists()).toBe(true);
  })

  it('status', () => {
    expect(renderTextarea(<Textarea status="success" />).classes()).toContain('i-textarea__inner--status-success');
  });

  it('noResize', () => {
    expect(renderTextarea(<Textarea noResize />).attributes('style')).toMatch('resize: none');
  });
});

