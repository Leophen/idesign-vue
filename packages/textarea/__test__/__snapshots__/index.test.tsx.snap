// Vitest Snapshot v1

exports[`Textarea 组件测试 > create 1`] = ``;

exports[`Textarea 组件测试 > defaultValue 1`] = ``;

exports[`Textarea 组件测试 > placeholder 1`] = ``;

exports[`Textarea 组件测试 > value 1`] = ``;
